### PR TITLE
Studio: Improve offline mode for Sync tab

### PR DESCRIPTION
**Pull request by katinthehatsite**
## Related issues

Closes https://github.com/Automattic/dotcom-forge/issues/9808

## Proposed Changes

This PR handles offline mode for the `Sync` tab.

**On the modal: if the connection goes out when the user has the modal open**:

Screenshot 2024-11-15 at 3 21 37 PM

**On the Sync tab itself**:

Screenshot 2024-11-15 at 3 21 53 PM

Screenshot 2024-11-15 at 3 22 00 PM

## Testing Instructions

* Pull the changes from this branch
* Start Studio app with `STUDIO_SITE_SYNC=true npm start`
* Navigate to the `Sync` tab that has some sites connected 
* Disable your internet connection (in my case, it was enough to disconnect from WiFi)
* Confirm that the buttons to `Create a new site` and `Connect site` are disabled
* Confirm that when you hover over `Pull` or `Push` buttons, they are disabled and you can see a tooltip with information
* Enable your internet connection
* Click on `Connect site` to open a modal
* Disable your internet connection
* Observe that you see the notice that lays over the content and does not allow select or connect sites
* Repeat the same steps on a Studio site that has no WP.com sites connected to it and confirm the buttons to connect a site or create a new site are disabled

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

**Comment by katinthehatsite** [comment]
Since we do not have the exact design for this, I have some open questions that I felt as comments on the PR. Feel free to make any suggestions!

cc @matt-west @wojtekn

File: src/components/sync-sites-modal-selector.tsx [hunk 1]
```diff
@@ -49,39 +52,49 @@ export function SyncSitesModalSelector( {
...
+				/>
 
-				{ ! isLoading && ! isEmpty && (
-					<ListSites
-						syncSites={ filteredSites }
-						selectedSiteId={ selectedSiteId }
-						onSelectSite={ setSelectedSiteId }
-					/>
+				{ isOffline && (
+					<div className="absolute inset-0 bg-white/80 z-10 flex items-center justify-center">
```
Comment (katinthehatsite) [inline src/components/sync-sites-modal-selector.tsx:93]: Not sure if this needs more or less transparency. I tried with something like `bg-white/90` and it seems like it is not transparent enough to me. Happy to adjust this though

File: src/components/sync-connected-sites.tsx [hunk 2]
```diff
@@ -193,27 +197,36 @@ export function SyncConnectedSites( {
...
-												<Button
-													variant="link"
-													className="!text-black hover:!text-a8c-blueberry"
-													disabled={ isAnySitePulling }
-												>
-													<Icon icon={ cloudUpload } />
-													{ __( 'Push' ) }
-												</Button>
-											</div>
+											<Tooltip
```
Comment (katinthehatsite) [inline src/components/sync-connected-sites.tsx]: I added one tooltip for these two buttons as they are too close too each other and are too small and making two tooltips seemed too crowded and excessive. Happy to change this though

File: src/components/connect-create-buttons.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+					className={ cx( ! isOffline && '!text-a8c-blueberry !shadow-a8c-blueberry' ) }
+					disabled={ isOffline }
+					aria-disabled={ isOffline }
+				>
+					{ __( 'Create new site' ) }
+					<ArrowIcon />
+				</Button>
+			</Tooltip>
+		</>
+	);
```
Comment (katinthehatsite) [inline src/components/connect-create-buttons.tsx:66]: I decided to extract these two buttons in the common component as we are using them in two places and the code is quite repetitive

File: src/components/sync-sites-modal-selector.tsx (hunk 1 above)
Comment (matt-west) [reply src/components/sync-sites-modal-selector.tsx:93]: This looks good to me. 👍

File: src/components/sync-connected-sites.tsx (hunk 2 above)
Comment (matt-west) [reply src/components/sync-connected-sites.tsx]: That makes sense. Good call @katinthehatsite

**Review by matt-west — APPROVED** [review]
Looks good to me @katinthehatsite. Thanks!

I spotted an unrelated issue with the alignment of the offline icon in the header bar. Logged here: https://github.com/Automattic/studio/issues/668

**Comment by ivan-ottinger** [comment]
> I spotted an unrelated issue with the alignment of the offline icon in the header bar. Logged here: https://github.com/Automattic/studio/issues/668

Thanks for catching this, Matt.

Turns out this is a regression introduced in https://github.com/Automattic/studio/pull/648. I will remove the `Toggle sidebar` tooltip until we have a better solution for tooltips in https://github.com/Automattic/dotcom-forge/issues/9841.

**Review by ivan-ottinger — APPROVED** [review]
The proposed changes look great and work as expected. 👌🏼🙂 

I did not observe any regressions in my tests.

I also appreciate the creation of a separate `ConnectCreateButtons` component.

---

**Update**: Notices just one smaller issue: https://github.com/Automattic/studio/pull/666#issuecomment-2485725294.

**Comment by ivan-ottinger** [comment]
I have noticed one minor issue: When the sidebar is closed, the tooltip on the `Connect site` button is not fully visible:

Screen Capture on 2024-11-19 at 14-23-24

We could consider changing the `Tooltip` position to the `bottom-start` for instance (`placement="bottom-start"`):

Screen Capture on 2024-11-19 at 15-10-18

It works well with RTL languages too:

Screen Capture on 2024-11-19 at 15-10-56

If we decide to go this route, it would be probably better to set the same position for the `Create new site` button as well - for consistency.

This also makes me realize we will need similar changes for other tooltips - e.g. `Toggle sidebar` to accommodate RTL (but that's not related to your PR of course). → Created a separate task: https://github.com/Automattic/dotcom-forge/issues/9896.

**Comment by katinthehatsite** [comment]
> We could consider changing the Tooltip position to the bottom-start for instance (placement="bottom-start"):

Nice catch! I adjusted this in https://github.com/Automattic/studio/pull/666/commits/ef4234c6523997bf16072333639fbe8361bba9d3 . I set it to `top-start` as `bottom-start` was affecting in a strange way the case when there is no connected site (see example below):

Screenshot 2024-11-19 at 8 49 39 PM

> This also makes me realize we will need similar changes for other tooltips - e.g. Toggle sidebar to accommodate RTL (but that's not related to your PR of course). → Created a separate task: https://github.com/Automattic/dotcom-forge/issues/9896.

Nice, thanks for adding the issue. I sort of ran into it when dealing with `Pull` and `Push` buttons for offline mode and using the `placement` prompt seemed to be the best approach to solve this.

**Pull request merged**